src/main/java/org/javarosa/core/model/TriggerableDag.java

```diff
@@ -270,144 +291,148 @@ public void copyItemsetAnswer(FormInstance mainInstance, EvaluationContext evalC
+        if (existingQuickTriggerable != null) {
+            // We found a quick triggerable wrapping a triggerable matching the provided one.
+            // Since Triggerable.equals() doesn't have into account contexts or targets, the
+            // safest thing to do would be to create a new quick triggerable for the incoming
+            // triggerable, but we could cover both triggerables if we compute the intersection
+            // between their respective context and reuse the one we've already stored.
+            existingQuickTriggerable.intersectContextWith(triggerable);
+            return existingQuickTriggerable.getTriggerable();
+        }
+
+        QuickTriggerable newQuickTriggerable = QuickTriggerable.of(triggerable);
+        allTriggerables.add(newQuickTriggerable);
 
-            return t;
+        // Build the triggerable per trigger index
+        Set<TreeReference> triggers = triggerable.getTriggers();
+        for (TreeReference trigger : triggers) {
+            if (!triggerablesPerTrigger.containsKey(trigger))
+                triggerablesPerTrigger.put(trigger, new HashSet<>());
+            triggerablesPerTrigger.get(trigger).add(newQuickTriggerable);
         }
+
+        return triggerable;
     }
 
     /**
      * Finalize the DAG associated with the form's triggered conditions. This
      * will create the appropriate ordering and dependencies to ensure the
      * conditions will be evaluated in the appropriate orders.
-     *
-     * @throws IllegalStateException - If the trigger ordering contains an illegal cycle and the
-     *                               triggers can't be laid out appropriately
      */
-    public void finalizeTriggerables(FormInstance mainInstance, EvaluationContext evalContext) throws IllegalStateException {
-        //
-        // DAGify the triggerables based on dependencies and sort them so that
-        // triggerables come only after the triggerables they depend on
-        //
-        List<QuickTriggerable> vertices = new ArrayList<>(unorderedTriggerables);
-        triggerablesDAG.clear();
-        List<QuickTriggerable[]> partialOrdering = new ArrayList<>();
-        Set<QuickTriggerable> newDestinationSet = new HashSet<>();
-        for (QuickTriggerable qt : vertices) {
-            Set<QuickTriggerable> deps = new HashSet<>();
-            newDestinationSet.clear();
-            fillTriggeredElements(mainInstance, evalContext, qt, deps, newDestinationSet);
-
-            if (deps.contains(qt))
-                throwCyclesInDagException(deps);
-
-            if (qt.canCascade())
-                for (QuickTriggerable qu : deps) {
-                    QuickTriggerable[] edge = {qt, qu};
-                    partialOrdering.add(edge);
-                }
+    public void finalizeTriggerables(FormInstance mainInstance) throws IllegalStateException {
+        List<QuickTriggerable[]> edges = getDagEdges(allTriggerables, triggerablesPerTrigger);
```

Comment:
Currently this is only ever used when all the contents are put in a set so there's no reason for this to be a list. 

It occurs to me that there are no tests that fail if the dag is unsorted. That is, I can go from `LinkedHashSet` to `HashSet` in the initialization of `triggerablesDAG` and all tests still pass. The `buildDag` algorithm is now easy to understand (I wish I could quickly generate an animation because it's very cool how the ordering is guaranteed) so that gives confidence but test coverage would would help me sleep even better at night. Does this seem worthwhile, @ggalmazor?

I'm not entirely sure what such a test would look like. I'd want to do something like force a hash code for an element at the end of a triggerable chain to sort at the beginning of `triggerablesDAG`.

Reply:
> Currently this is only ever used when all the contents are put in a set so there's no reason for this to be a list. 

Let's make it a `Set`. It's an inexpensive change and it makes the context richer.

> It occurs to me that there are no tests that fail if the dag is unsorted [...] I'm not entirely sure what such a test would look like

I have two ideas:
- We could write a couple of simple low-level tests around `buildDag` that explicitly show how the output DAG is sorted while describing its behavior.
- We could write an end-to-end test with 3 fields `a`, `b`, `c`. 
  - `b` depends on `a` as in `a * 2`.
  - `c` depends on `a` and `b`, as in `a * b`.
  - We give initial values to all fields in the form.
  - We change `a` to trigger all computations and verify that `c` was computed using the updated value of `b`, not the initial value in the form.

Reply:
I've pushed a commit including the end-to-end test, which I think would be enough. Let me know if you'd want the low-level test too.

Reply:
> and it doesn't guarantee the order of its elements

Well, it guarantees that they'll be in the insertion order. But no matter, we agree on the change to `Set` which is great.

> I've pushed a commit including the end-to-end test, which I think would be enough.

Ah yes, perfect!

Reply:
> Well, it guarantees that they'll be in the insertion order. But no matter, we agree on the change to Set which is great.

Ouch, you're right! I guess I still have the parallel streams mindset, where you can't trust the order will continue being the same as the original insertion order after some stream manipulations.